pkg/cli/mirror/copy.go

```diff
@@ -412,42 +472,109 @@ func getConfigPathFromConfigLayer(imagePath, configSha string) (string, error) {
 	return "", fmt.Errorf("label %s not found in config blob %s", configsLabel, configLayerDir)
 }
 
-// getRelatedImages this reads each catalog or config.json
-// file in a given operator in the FBC
-// and returns the list of relatedImages found in the CSV
-func getRelatedImages(path string) ([]model.RelatedImage, error) {
-	var icJSON *model.Bundle
-
-	// read the config.json to get releated images
-	icd, err := ioutil.ReadFile(path + catalogJSON)
+// getRelatedImages reads a directory containing an FBC catalog () unpacked contents
+// and returns the list of relatedImages found in the CSVs of bundles
+// filtering by the list of packages provided in imageSetConfig for the catalog
+func getRelatedImages(directory string, packages []v1alpha2.IncludePackage) ([]declcfg.RelatedImage, error) {
+	allImages := []declcfg.RelatedImage{}
+	// load the declarative config from the provided directory (if possible)
+	cfg, err := declcfg.LoadFS(os.DirFS(directory))
 	if err != nil {
 		return nil, err
 	}
-	// we are only interested in the related images section
-	tmp := string(icd)
-	i := strings.Index(tmp, relatedImages)
-	newJson := "{\"" + tmp[i:]
-	err = json.Unmarshal([]byte(newJson), &icJSON)
-	if err != nil {
-		return nil, fmt.Errorf("json unmarshal %v", err)
+
+	if len(packages) == 0 {
+		for _, aPackage := range cfg.Packages {
+			packages = append(packages, v1alpha2.IncludePackage{
+				Name: aPackage.Name,
+			})
+		}
 	}
-	//unique images - remove duplicate values
-	images := []model.RelatedImage{}
 
-	for _, v := range icJSON.RelatedImages {
+	for _, bundle := range cfg.Bundles {
+		isSelected, err := isPackageSelected(bundle, cfg.Channels, packages)
+		if err != nil {
+			return nil, err
+		}
+		if isSelected {
+			allImages = append(allImages, declcfg.RelatedImage{Name: bundle.Package, Image: bundle.Image})
+			allImages = append(allImages, bundle.RelatedImages...)
+		}
+	}
+	//make sure there are no duplicates in the list with same image:
+	finalList := []declcfg.RelatedImage{}
+	for _, i := range allImages {
 		found := false
-		for _, w := range images {
-			if w.Image == v.Image {
+		for _, j := range finalList {
+			if i.Image == j.Image {
 				found = true
 				break
 			}
 		}
 		if !found {
-			images = append(images, v)
+			finalList = append(finalList, i)
 		}
 	}
+	return finalList, nil
+}
 
-	return images, nil
+func isPackageSelected(bundle declcfg.Bundle, channels []declcfg.Channel, packages []v1alpha2.IncludePackage) (bool, error) {
+	isSelected := false
+	for _, pkg := range packages {
+		if pkg.Name == bundle.Package {
```

Comment:
nit: if you replace this condition with the following:
```golang
if pkg.Name != bundle.Package {
    // no filtering required
    isSelected = true
}
var min, max semver.Version
...
```
this make the code easier to read, because you have less indentation